### PR TITLE
Silence compiler warnings

### LINKED ISSUE #1799: format mismatch warnings
```
/home/joerg/src/avrdude/src/dfu.c:398:5: warning: format specifies type 'unsigned short' but the argument has type 'int' [-Wformat]
    ((unsigned short) dfu->dev_desc.bcdDevice >> 8) & 0xFF,
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/joerg/src/avrdude/src/avrdude.h:59:97: note: expanded from macro 'msg_info'
#define msg_info(...)       avrdude_message2(stderr, __LINE__, __FILE__, __func__, 0, MSG_INFO, __VA_ARGS__)
                                                                                                ^~~~~~~~~~~
/home/joerg/src/avrdude/src/dfu.c:399:5: warning: format specifies type 'unsigned short' but the argument has type 'int' [-Wformat]
    ((unsigned short) dfu->dev_desc.bcdDevice >> 4) & 0xF,
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/joerg/src/avrdude/src/avrdude.h:59:97: note: expanded from macro 'msg_info'
#define msg_info(...)       avrdude_message2(stderr, __LINE__, __FILE__, __func__, 0, MSG_INFO, __VA_ARGS__)
                                                                                                ^~~~~~~~~~~
/home/joerg/src/avrdude/src/dfu.c:400:5: warning: format specifies type 'unsigned short' but the argument has type 'int' [-Wformat]
    ((unsigned short) dfu->dev_desc.bcdDevice >> 0) & 0xF);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/joerg/src/avrdude/src/avrdude.h:59:97: note: expanded from macro 'msg_info'
#define msg_info(...)       avrdude_message2(stderr, __LINE__, __FILE__, __func__, 0, MSG_INFO, __VA_ARGS__)
…
/home/joerg/src/avrdude/src/flip2.c:559:5: warning: format specifies type 'unsigned short' but the argument has type 'int' [-Wformat]
    ((unsigned short) flip2->boot_ver >> 4) & 0xF,
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/joerg/src/avrdude/src/avrdude.h:59:97: note: expanded from macro 'msg_info'
#define msg_info(...)       avrdude_message2(stderr, __LINE__, __FILE__, __func__, 0, MSG_INFO, __VA_ARGS__)
                                                                                                ^~~~~~~~~~~
/home/joerg/src/avrdude/src/flip2.c:560:5: warning: format specifies type 'unsigned short' but the argument has type 'int' [-Wformat]
    ((unsigned short) flip2->boot_ver >> 0) & 0xF);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/joerg/src/avrdude/src/avrdude.h:59:97: note: expanded from macro 'msg_info'
#define msg_info(...)       avrdude_message2(stderr, __LINE__, __FILE__, __func__, 0, MSG_INFO, __VA_ARGS__)
                                                                                                ^~~~~~~~~~~
/home/joerg/src/avrdude/src/flip2.c:781:70: warning: format specifies type 'unsigned short' but the argument has type 'int' [-Wformat]
      pmsg_error("address out of range [0x%04hX,0x%04hX]\n", offset, offset+size-1);
                                                  ~~~~~              ^~~~~~~~~~~~~
                                                  %04X
/home/joerg/src/avrdude/src/avrdude.h:67:176: note: expanded from macro 'pmsg_error'
#define pmsg_error(...)     avrdude_message2(stderr, __LINE__, __FILE__, __func__, MSG2_PROGNAME|MSG2_FUNCTION|MSG2_FILELINE|MSG2_TYPE|MSG2_FLUSH|MSG2_LEFT_MARGIN, MSG_ERROR, __VA_ARGS__)
                                                                                                                                                                               ^~~~~~~~~~~
/home/joerg/src/avrdude/src/flip2.c:839:70: warning: format specifies type 'unsigned short' but the argument has type 'int' [-Wformat]
      pmsg_error("address out of range [0x%04hX,0x%04hX]\n", offset, offset+size-1);
                                                  ~~~~~              ^~~~~~~~~~~~~
                                                  %04X
/home/joerg/src/avrdude/src/avrdude.h:67:176: note: expanded from macro 'pmsg_error'
#define pmsg_error(...)     avrdude_message2(stderr, __LINE__, __FILE__, __func__, MSG2_PROGNAME|MSG2_FUNCTION|MSG2_FILELINE|MSG2_TYPE|MSG2_FLUSH|MSG2_LEFT_MARGIN, MSG_ERROR, __VA_ARGS__)
                                                                                                                                                                               ^~~~~~~~~~~
```
There would at least be an `(unsigned short)` typecast missing.

Anyway, I wonder what's the point of forcing all that into unsigned short, rather than keeping it simply `unsigned`, and remove all the `h` format modificators?

### PR DESCRIPTION
**Pull request by stefanrueger**
Fixes #1799 

The `%hd` converts the int argument to short before printing, so I suspect #1799 is a matter of overzealous compiler warning.

### CODE REVIEW

**Comment by MCUdude** [comment]
Regarding compiler warnings, I'm getting the following one when building git main. No big deal, but I thought I'd mention it.

```
  LEX      lexer.c
/Library/Developer/CommandLineTools/usr/bin/make  all-recursive
Making all in .
  CC       avrdude-main.o
  CC       avrdude-whereami.o
  CC       avrdude-developer_opts.o
  CC       libavrdude_a-config_gram.o
  CC       libavrdude_a-lexer.o
lexer.c:2434:38: warning: comparison of integers of different signs: 'unsigned long' and 'int' [-Wsign-compare]
        if (((yy_n_chars) + number_to_move) > YY_CURRENT_BUFFER_LVALUE->yy_buf_size) {
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

**Comment by stefanrueger** [comment]
@MCUdude That's a known issue and we cannot do anything about it, as it's MACOS's flex that generates the code.

File: src/dfu.c [hunk 1]
```diff
@@ -394,10 +394,10 @@ void dfu_show_info(struct dfu_dev *dfu)
     msg_info("    USB Product         : 0x%04hX\n",
```
Comment (dl8dtl) [inline src/dfu.c:394]: Shouldn't we also remove the `h` and `(unsigned short)` cast here?

Comment (stefanrueger) [reply src/dfu.c:394]: No, that would change what happens. Right now printf's %hx will (unsigned short) the promoted int argument that itself already has been (unsigned short)-ed. We can remove either the `h` or the `(unsigned short)` cast of the (assuming at most int-wide) argument and the print will be the same. If we remove both then if the arg doesn't fit in an unsigned short something else will be printed.

Comment (stefanrueger) [reply src/dfu.c:394]: Here an example:
```
#include <stdio.h>
int main() {
  int i = 0x78000;
  printf("%04hx\n", (unsigned short) i);
  printf("%04x\n", (unsigned short) i);
  printf("%04hx\n", i);
  printf("%04x\n", i);
}
```

**Comment by mcuee** [comment]
MinGW build still has some warnings.

```
[ 68%] Building C object src/CMakeFiles/libavrdude.dir/updi_link.c.obj
In file included from D:/a/avrdude/avrdude/src/updi_link.c:36:
D:/a/avrdude/avrdude/src/updi_link.c: In function 'updi_physical_send':
D:/a/avrdude/avrdude/src/updi_link.c:93:14: warning: format '%lu' expects argument of type 'long unsigned int', but argument 8 has type 'size_t' {aka 'long long unsigned int'} [-Wformat=]
   93 |   pmsg_debug("sending %lu bytes [", len);
      |              ^~~~~~~~~~~~~~~~~~~~~  ~~~
      |                                     |
      |                                     size_t {aka long long unsigned int}
D:/a/avrdude/avrdude/src/avrdude.h:91:138: note: in definition of macro 'pmsg_debug'
   91 | #define pmsg_debug(...)     avrdude_message2(stderr, __LINE__, __FILE__, __func__, MSG2_PROGNAME|MSG2_FLUSH|MSG2_LEFT_MARGIN, MSG_DEBUG, __VA_ARGS__)
      |                                                                                                                                          ^~~~~~~~~~~
D:/a/avrdude/avrdude/src/updi_link.c:93:25: note: format string is defined here
   93 |   pmsg_debug("sending %lu bytes [", len);
      |                       ~~^
      |                         |
      |                         long unsigned int
      |                       %llu
D:/a/avrdude/avrdude/src/updi_link.c: In function 'updi_physical_recv':
D:/a/avrdude/avrdude/src/updi_link.c:117:14: warning: format '%lu' expects argument of type 'long unsigned int', but argument 8 has type 'size_t' {aka 'long long unsigned int'} [-Wformat=]
  117 |   pmsg_debug("received %lu bytes [", len);
      |              ^~~~~~~~~~~~~~~~~~~~~~  ~~~
      |                                      |
      |                                      size_t {aka long long unsigned int}
D:/a/avrdude/avrdude/src/avrdude.h:[91](https://github.com/avrdudes/avrdude/actions/runs/8870239966/job/24351804351#step:5:92):138: note: in definition of macro 'pmsg_debug'
   91 | #define pmsg_debug(...)     avrdude_message2(stderr, __LINE__, __FILE__, __func__, MSG2_PROGNAME|MSG2_FLUSH|MSG2_LEFT_MARGIN, MSG_DEBUG, __VA_ARGS__)
      |                                                                                                                                          ^~~~~~~~~~~
D:/a/avrdude/avrdude/src/updi_link.c:117:26: note: format string is defined here
  117 |   pmsg_debug("received %lu bytes [", len);
      |                        ~~^
      |                          |
      |                          long unsigned int
      |                        %llu
```

**Comment by stefanrueger** [comment]
> `size_t` requires format letter `z`

Correct; this was introduced with C99, but we have in the past tried to be even compatible with pre-1999 C-libraries (#1608) so probably better to cast `len` to `(unsigned long)` in this case. Just pushed a commit for this

**Comment by dl8dtl** [comment]
I wouldn't do so unless MSVC is still that far behind to require it.
We're requiring a lot of C99 already elsewhere anyway.

**Comment by dl8dtl** [comment]
Quite sad, it's only 15 years that have gone since …

**Comment by mcuee** [comment]
@stefanrueger and @dl8dtl 

My suggestion is to give up on that paticular configuration -- mingw32 cross compiler toolchain under Windows 7.

If you read my comments in #1594, there is no issue using MSVC build and normal MSYS2 mingw32/64 build.

**Comment by mcuee** [comment]
> I wouldn't do so unless MSVC is still that far behind to require it. We're requiring a lot of C99 already elsewhere anyway.

MSVC is not the problem here. So one less worry.

**Comment by mcuee** [comment]
@dl8dtl and @stefanrueger and @MCUdude 

> Quite sad, it's only 15 years that have gone since …

Let's face it, Windows 7 was released on 22-Sept-2009 and Microsoft stopped the extended support on 10-Jan-2023 (main line support ended on 14-Jan-2020). I think it is perfectly fine that we drop the support of Windows 7 altogether.

I do not have Windows 7 machine and I am not so sure if anyone of use still wants to carry out test under Windows 7.

**Comment by dl8dtl** [comment]
You're right about Win7. But then, I think it's fine to use the 'z' modifier, rather than casting stuff around.

**Comment by mcuee** [comment]
> You're right about Win7. But then, I think it's fine to use the 'z' modifier, rather than casting stuff around.

No issue with that. I just want to establish that we will not officially support Windows 7/8/8.1 from now on. It may still work but we will rely the users to provide non-intrusive patches if they really care about Windows 7/8/8.1.

BTW, MSYS2 has dropped 32bit Windows support of libusb/libusb-compat/libftdi, so we have already dropped 32bit mingw support of avrdude as well. Take note Windows 10/11 only supports x86_64 and not 32bit x86.

**Comment by mcuee** [comment]
@stefanrueger and @dl8dtl 

Not sure if we can fix MSYS2 clang64 compiler warnings as well.
```
D:/a/avrdude/avrdude/src/ser_win32.c:436:3: warning: comparison of integers of different signs: 'SOCKET' (aka 'unsigned long long') and 'const int' [-Wsign-compare]
  436 |                 FD_SET(fd->ifd, &rfds);
      |                 ^~~~~~~~~~~~~~~~~~~~~~
D:/a/_temp/msys64/clang64/include/psdk_inc/_fd_types.h:77:40: note: expanded from macro 'FD_SET'
   77 |                 if (((fd_set *)(set))->fd_array[__i] == (fd)) {         \
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^   ~~
D:/a/avrdude/avrdude/src/ser_win32.c:[55](https://github.com/avrdudes/avrdude/actions/runs/8935908735/job/24545265227#step:5:56)7:3: warning: comparison of integers of different signs: 'SOCKET' (aka 'unsigned long long') and 'const int' [-Wsign-compare]
  557 |                 FD_SET(fd->ifd, &rfds);
      |                 ^~~~~~~~~~~~~~~~~~~~~~
D:/a/_temp/msys64/clang64/include/psdk_inc/_fd_types.h:77:40: note: expanded from macro 'FD_SET'
   77 |                 if (((fd_set *)(set))->fd_array[__i] == (fd)) {         \
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^   ~~
2 warnings generated.
```

**Comment by stefanrueger** [comment]
> Not sure if we can fix MSYS2 clang64 compiler warnings as well.

See [here](https://github.com/avrdudes/avrdude/issues/1541#issuecomment-1783937965)

**Pull request merged**